TombEngine/Objects/TR1/Trap/ElectricBall.cpp

```diff
@@ -0,0 +1,463 @@
+
+			 if (item.Model.Color.z <= EldectricBallData.Colorw.z)
+				 item.Model.Color.z = EldectricBallData.Colorw.z;
+			 else
+				 item.Model.Color.z -= 0.1f;
+		}
+		
+		if (item.ItemFlags[3] > 28)
+		{
+			SpawnElectricityGlow(origin.ToVector3(), 108, 22, 22, 54);
+		}
+		else
+		{			
+			SpawnElectricityGlow(origin.ToVector3(), 128, 102, 102, 204);		
+			SpawnElectricity(origin.ToVector3() - Vector3(0, 1300, 0), origin.ToVector3() - (targetRandom.ToVector3() * 8), 15, 32, 64, 128, 30, (int)ElectricityFlags::Spline | (int)(int)ElectricityFlags::ThinOut | (int)ElectricityFlags::MoveEnd, 8, 12);//4
+			SpawnElectricity(origin.ToVector3() - Vector3(0, 1300, 0), origin.ToVector3() - (targetRandom.ToVector3() * 8), 15, 32, 64, 128, 30, (int)ElectricityFlags::Spline | (int)(int)ElectricityFlags::ThinOut | (int)ElectricityFlags::MoveEnd, 8, 12);//4
+			SpawnElectricity(origin.ToVector3() + Vector3(0, 300, 0), origin.ToVector3() + (targetRandom.ToVector3() * 8), 15, 32, 64, 128, 30, (int)ElectricityFlags::Spline | (int)(int)ElectricityFlags::ThinOut | (int)ElectricityFlags::MoveEnd, 8, 12);//4
+			SpawnElectricity(origin.ToVector3() + Vector3(0, 300, 0), origin.ToVector3() + (targetRandom.ToVector3() * 8), 15, 32, 64, 128, 30, (int)ElectricityFlags::Spline | (int)(int)ElectricityFlags::ThinOut | (int)ElectricityFlags::MoveEnd, 8, 12);//4
+			SpawnElectricEffect(item, 0, Vector3i(0, 0, 0), BLOCK(0.45f), BLOCK(0.45f), BLOCK(0.25f), 80, Vector3::Zero);
+		}
+		
+		if (item.ItemFlags[6] > 0)
+		{
+			if (ObjectOnLOS2(&origin, &target, &hitPos, nullptr, ID_LARA) == LaraItem->Index)
+			{
+				if (LaraItem->HitPoints <= ELECTRIC_BALL_LIGHTNING_DAMAGE)
+				{
+					ItemElectricBurn(LaraItem);
+					DoDamage(LaraItem, ELECTRIC_BALL_LIGHTNING_DAMAGE);
+				}
+				else
+				{
+					DoDamage(LaraItem, ELECTRIC_BALL_LIGHTNING_DAMAGE);
+				}			
+			}
+			item.ItemFlags[6]--;
+			SoundEffect(SFX_TR5_GOD_HEAD_LASER_LOOPS, &item.Pose);
+			SoundEffect(SFX_TR4_ELECTRIC_ARCING_LOOP, &item.Pose);
+		}	
+	}
+
+	void ControlElectricBall(short itemNumber)
+	{
+		if (!CreatureActive(itemNumber))
+			return;
+
+		auto& item = g_Level.Items[itemNumber];
+
+		auto& object = Objects[item.ObjectNumber];
+		const auto& laraItem = *LaraItem;
+		static auto targetPos = Vector3i::Zero;
+
+		AI_INFO ai;
+		CreatureAIInfo(&item, &ai);
+
+		if (ai.distance > SQUARE(BLOCK(7)) )
+		{
+			item.ItemFlags[3] = 70;
+			item.ItemFlags[4] = 70;
+			item.Model.Color = EldectricBallData.Colorw;
+			item.ItemFlags[1] = TargetType::None;
+			return;
+		}
+
+		SoundEffect(SFX_TR5_DOOR_BEAM, &item.Pose);
+
+		auto targetPlayer = GameVector(GetJointPosition(LaraItem, LM_TORSO), LaraItem->RoomNumber);;
+		auto origin = GameVector(GetJointPosition(item, 0), item.RoomNumber);
+		bool los = LOS(&origin, &targetPlayer);
+		FloorInfo* floor = GetFloor(item.Pose.Position.x, item.Pose.Position.y, item.Pose.Position.z, &item.RoomNumber);
+
+		if (ai.distance < SQUARE(BLOCK(3)) && los)
```

Comment:
You could use the OCB as attack range ? OCB 1 = SQUARE(BLOCK(1))
or use OCB for impact point range detection ?

Reply:
Some complained that the range of 3 is too small, OCB should fix it :)

Reply:
Or you can use ItemFlags and make an index a IMPACT detection or LARA range detection range, and can be changed through LUA ?